prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -1230,3 +1251,244 @@ void rxr_pkt_handle_rtr_recv(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	rx_entry->state = RXR_RX_WAIT_READ_FINISH;
 	rxr_pkt_entry_release_rx(ep, pkt_entry);
 }
+
+/*
+ * RTA packet functions
+ */
+ssize_t rxr_pkt_init_rta(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
+			 int pkt_type, struct rxr_pkt_entry *pkt_entry)
+{
+	struct fi_rma_iov *rma_iov;
+	struct rxr_rta_hdr *rta_hdr;
+	char *data;
+	size_t data_size;
+	int i;
+
+	rta_hdr = (struct rxr_rta_hdr *)pkt_entry->pkt;
+	rta_hdr->msg_id = tx_entry->msg_id;
+	rta_hdr->rma_iov_count = tx_entry->rma_iov_count;
+	rta_hdr->atomic_datatype = tx_entry->atomic_hdr.datatype;
+	rta_hdr->atomic_op = tx_entry->atomic_hdr.atomic_op;
+	rta_hdr->tx_id = tx_entry->tx_id;
+	rxr_pkt_init_req_hdr(ep, tx_entry, pkt_type, pkt_entry);
+	rta_hdr->flags |= RXR_REQ_ATOMIC;
+	rma_iov = rta_hdr->rma_iov;
+	for (i=0; i < tx_entry->rma_iov_count; ++i) {
+		rma_iov[i].addr = tx_entry->rma_iov[i].addr;
+		rma_iov[i].len = tx_entry->rma_iov[i].len;
+		rma_iov[i].key = tx_entry->rma_iov[i].key;
+	}
+
+	data = (char *)pkt_entry->pkt + pkt_entry->hdr_size;
+	data_size = ofi_copy_from_iov(data, ep->mtu_size - pkt_entry->hdr_size,
+				      tx_entry->iov, tx_entry->iov_count, 0);
+
+	pkt_entry->pkt_size = pkt_entry->hdr_size + data_size;
+	pkt_entry->x_entry = tx_entry;
+	return 0;
+}
+
+ssize_t rxr_pkt_init_write_rta(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
+			    struct rxr_pkt_entry *pkt_entry)
+{
+	rxr_pkt_init_rta(ep, tx_entry, RXR_WRITE_RTA_PKT, pkt_entry);
+	return 0;
+}
+
+ssize_t rxr_pkt_init_fetch_rta(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
+			      struct rxr_pkt_entry *pkt_entry)
+{
+	rxr_pkt_init_rta(ep, tx_entry, RXR_FETCH_RTA_PKT, pkt_entry);
+	return 0;
+}
+
+ssize_t rxr_pkt_init_compare_rta(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
+				 struct rxr_pkt_entry *pkt_entry)
+{
+	char *data;
+	size_t data_size;
+
+	rxr_pkt_init_rta(ep, tx_entry, RXR_COMPARE_RTA_PKT, pkt_entry);
+
+	/* rxr_pkt_init_rta() will copy data from tx_entry->iov to pkt entry
+	 * the following append the data to be compared
+	 */
+	data = (char *)pkt_entry->pkt + pkt_entry->pkt_size;
+	data_size = ofi_copy_from_iov(data, ep->mtu_size - pkt_entry->pkt_size,
+				      tx_entry->atomic_ex.comp_iov,
+				      tx_entry->atomic_ex.comp_iov_count, 0);
+	assert(data_size == tx_entry->total_len);
+	pkt_entry->pkt_size += data_size;
+	return 0;
+}
+
+void rxr_pkt_handle_write_rta_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_tx_entry *tx_entry;
+
+	tx_entry = (struct rxr_tx_entry *)pkt_entry->x_entry;
+	rxr_cq_handle_tx_completion(ep, tx_entry);
+}
+
+void rxr_pkt_handle_fetch_rta_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
```

Comment:
If we have no action to take, why call an empty function?

Reply:
Removed the function and added comments in rxr_pkt_cmd.c.